### PR TITLE
[pkg/ottl] Add grammar utility to extract paths from statements

### PR DESCRIPTION
**Pull request by edmocosta**
**Description:** <Describe what has changed.>

This PR is part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017, and adds a grammar utility to allow extracting `ottl.path` from parsed statements and expressions. The new functions will power the context inferrer utility (see [draft](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050)), which implementation will basically gather all declared statement's `Path.Context`, and choose the highest priority one.

For the statements rewriter utility purpose, it changes the `grammar.go` file including a new field `Pos lexer.Position` into the `ottl.path` struct. This value is automatically set by the `participle` lexer and holds the path's offsets, being useful for identifying their positions in the raw statement, without using regexes an being coupled to the grammar's definition.

**Additional changes**:

This proposal uses the visitor pattern to gather all path's from the parsed statements. Considering the grammar's custom error mechanism (`checkForCustomError`) also works with a similar pattern, I've added the visitor implementation as part of the grammar objects, and refactored all `checkForCustomError` functions to be part of a validator visitor. Differently of the current implementation, it joins and displays all errors at once instead of one by one, IMO, it's specially useful for statements with more than one error, for example: `set(name, int(2), float(1))`.

*&ast;* We can change it back do be error-by-error if necessary.
*&ast;* If modifying the custom validator is not desired, I can roll those changes back keeping only the necessary code to extract the path's + `Pos` field.

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017

**Testing:** Unit tests were added

**Documentation:** No changes

### CODE REVIEW

**Comment by TylerHelmuth** [comment]
@edmocosta I'm holding off on reviewing this until it is marked Ready, but I am interested.

**Comment by edmocosta** [comment]
> @edmocosta I'm holding off on reviewing this until it is marked Ready, but I am interested.

I was running a few more tests and forgot to mark it as ready. Thanks @TylerHelmuth!

File: pkg/ottl/grammar.go
Comment (TylerHelmuth) [reply pkg/ottl/grammar.go:267]: Where does it get used? Is it testing only? If so thats fine.

File: pkg/ottl/grammar.go
Comment (TylerHelmuth) [inline pkg/ottl/grammar.go:497]: @edmocosta as I'm reading this PR I feel like I'm right on the edge of full comprehension, but I think I'm getting tripped up by the combining of `checkForCustomError` and checking all the places we visit.  

I know it'll mean traversing the parsed structs twice, but can you separate the new feature from the error checking?

Separately, I like how you've updated the error checking to combine the errors instead of returning the first one found, that is a nice quality of life change and is worthy of its own PR.

Comment (edmocosta) [reply pkg/ottl/grammar.go:497]: Sure thing, I will move it to a separate PR.

File: pkg/ottl/grammar.go
Comment (edmocosta) [reply pkg/ottl/grammar.go:267]: It's not used yet, but the idea was to use it to find all statement's `path`'s offsets, without the need to use regexes.
Given we can have very complex statements, this option, IMO, would be much safer, efficient, and future proof, as it's part of the grammar and does not depend on regexes or any other logic to find where a path starts. 

There's an example on how I was planning to use it on the context rewriter [draft](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050/files#diff-03057e9428362a42cca8c0e18914753e886b06a57ac8bafaa7ea4d42504d02f9R204).

File: pkg/ottl/grammar.go
Comment (edmocosta) [reply pkg/ottl/grammar.go:624]: Done, I'll be creating the `checkForCustomError` PR as soon as we merge this one. Thanks!

Comment (TylerHelmuth) [reply pkg/ottl/grammar.go:624]: I like this better thank you

**Review by TylerHelmuth — APPROVED** [review]

File: .chloggen/ottl_add_grammar_paths_visitor.yaml
Comment (TylerHelmuth) [inline .chloggen/ottl_add_grammar_paths_visitor.yaml:1]: We can remove this changelog since the PR has no end-user or API changes.

Comment (edmocosta) [reply .chloggen/ottl_add_grammar_paths_visitor.yaml:1]: Done! thanks for reviewing!

**Comment by edmocosta** [comment]
Hi @TylerHelmuth! please let me know if there is anything left before it can be merged. Thank you!

**Review by evan-bradley — APPROVED** [review]
Looks good to me, sorry for the long delay in reviewing this. Thank you @edmocosta!

**Pull request merged**
